components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java

```diff
@@ -0,0 +1,332 @@
+import org.wso2.carbon.apimgt.impl.AbstractKeyManager;
+import org.wso2.carbon.apimgt.impl.kmclient.KMClientErrorDecoder;
+import org.wso2.carbon.apimgt.impl.kmclient.KeyManagerClientException;
+import org.wso2.carbon.apimgt.impl.recommendationmgt.AccessTokenGenerator;
+
+import feign.Feign;
+import feign.Feign.Builder;
+import feign.gson.GsonDecoder;
+import feign.gson.GsonEncoder;
+import feign.okhttp.OkHttpClient;
+import feign.slf4j.Slf4jLogger;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class AzureADClient extends AbstractKeyManager {
+    /***
+     * Application related API calls
+     */
+    private ApplicationClient appClient;
+
+    private String tokenEndpoint;
+
+    @Override
+    public void loadConfiguration(KeyManagerConfiguration configuration) throws APIManagementException {
+        this.configuration = configuration;
+
+        String appClientId = (String) this.configuration.getParameter(AzureADConstants.AD_APP_CLIENT_ID);
+        String appClientSecret = (String) this.configuration.getParameter(AzureADConstants.AD_APP_CLIENT_SECRET);
+        String revokeEndpoint = (String) this.configuration.getParameter(APIConstants.KeyManager.REVOKE_ENDPOINT);
+        String graphApiEndpoint = (String) this.configuration.getParameter(AzureADConstants.GRAPH_API_ENDPOINT);
+
+        tokenEndpoint = (String) this.configuration.getParameter(APIConstants.KeyManager.TOKEN_ENDPOINT);
+
+        AccessTokenGenerator accessTokenGenerator = new AccessTokenGenerator(tokenEndpoint, revokeEndpoint, appClientId,
+                appClientSecret);
+
+        AzureADRequestInterceptor appInterceptor = new AzureADRequestInterceptor(accessTokenGenerator);
+
+        appClient = this.buildFeignClient(new OkHttpClient(), appInterceptor)
+                .target(ApplicationClient.class, graphApiEndpoint);
+    }
+
+    private Builder buildFeignClient(
+            OkHttpClient client, AzureADRequestInterceptor interceptor) {
+        return Feign.builder()
+                .client(client)
+                .requestInterceptor(interceptor)
+                .encoder(new GsonEncoder())
+                .decoder(new GsonDecoder())
+                .errorDecoder(new KMClientErrorDecoder())
+                .logger(new Slf4jLogger());
+    }
+
+    @Override
+    public OAuthApplicationInfo createApplication(OAuthAppRequest oauthAppRequest) throws APIManagementException {
+        OAuthApplicationInfo oauthAppInfo = oauthAppRequest.getOAuthApplicationInfo();
+
+        if (oauthAppInfo != null) {
+            ClientInformation appInfo = this.getClientInformation(oauthAppInfo);
+            ClientInformation app;
+
+            try {
+                app = appClient.createApplication(appInfo);
+                if (app != null) {
+                    PasswordInfo pInfo = this.setPassword(app.getId());
+                    app.setClientSecret(pInfo.getSecret());
+                    return getOAuthApplicationInfo(app);
+                }
```

Comment:
Is the else block of this "if" an error situation?
Should we handle this case too?

Reply:
handled with an else and throw an exception